python/ray/autoscaler/_private/util.py

```diff
@@ -724,6 +724,18 @@ def filter_placement_group_from_bundle(bundle: ResourceBundle):
     return demand_lines
 
 
+def get_constraints_report(lm_summary: LoadMetricsSummary):
+    constraint_lines = []
+    for bundle, count in lm_summary.request_demand:
+        line = f" {bundle}: {count} from request_resources()"
+        constraint_lines.append(line)
+    if len(constraint_lines) > 0:
+        constraints_report = "\n".join(constraint_lines)
+    else:
+        constraints_report = " (no request_resources() constraints)"
+    return constraints_report
+
+
 def get_demand_report(lm_summary: LoadMetricsSummary):
     demand_lines = []
     if lm_summary.resource_demand:
```

Comment:
Can you help me understand the difference between

`demand_lines.extend(format_resource_demand_summary(lm_summary.resource_demand))` (L730) and `lm_summary.request_demand` (L736)?

Reply:
`resource_demand` consists of `ready/infeasible/blacklog` task shapes for autoscaler to scale the cluster.
`request_demand` is made by a direct `request_resources()` call from the user.